README
This is a "plugin" for the Video Disk Recorder (VDR).

Written by:                  Denis Loh <denis.loh@hftl.de>

Project's homepage:          http://upnp.vdr-developer.org

Latest version available at: http://upnp.vdr-developer.org

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
See the file COPYING for more information.

Please note: This plugin is written under the terms of open source
redistribution by

Hochschule für Telekommunikation Leipzig,
University of Applied Science
Gustav-Freytag-Straße 43-45
04277 Leipzig
Germany
http://www.hft-leipzig.de

All rights reserved.

Contents:

1 Description
2 Dependencies
3 Installation
4 Firewall

Description:

UPnP/DLNA Plugin for Video Disk Recorder

This Plugins extends the VDR with the possibility to act as an UPnP/DLNA Media
Server (DMS). It will serve VDR's contents in the network to any UPnP-AV and
DLNA capable devices.

The plugin requires a valid network connection with a IPv4 address. It will find
its settings automatically by default. In this case, the first network device in
the network device list with a valid IPv4 address and port 49152 will be used
for connections. You can change this behavior either by setting the command line
options or by editing the setup via the VDRs setup page.

Dependencies:

This plugin is tested with and requires the following libraries to work:

libupnp-1.6.6               The UPnP SDK
libsqlite-3.6               SQLite 3 Database
libtntnet-2.0               TNTnet library web server
libtntdb-1.1.1              TNTdb library database connector

boost::spirit               Grammar parser
boost::function             Boost function library
boost::bind                 Boost bind library

The boost libraries are usually available in a packed version, so that only a
single installation is required.

Installation:

To install the plugin and its subplugins you just have to

make
make subplugins

If you want to have icons when you open the server from a client, you have to
copy the httpdocs folder to the configuration directory of the plugin, for
instance:

cp httpdocs /var/lib/vdr/plugins/upnp/

I appreciate, if you help making the plugin better. If you find any project
related errors or missing features, please open a new ticket at the plugin
website with a detailed description of what you want or what happened.

Firewall:

If you want to use the plugin behind a firewall, then you need to open the
following ports:

1900            UDP     SSDP (Simple Service Discorvery Protocol)
2869            TCP     SSDP (Simple Service Discorvery Protocol)
7649            TCP		Web server default port
3000            TCP     Streamdev default port
8001            TCP     Live default port
User defined    TCP     Data

The user defined port may be set automatically by the server or by the user via
the setup or command line options. You need to accept multicast messages from
239.255.255.250, which is used by SSDP to discover other UPnP devices.


HISTORY
VDR Plugin 'upnp' Revision History
----------------------------------

2013-01-03: Version 1.0.0

 - Fix: Fixed many UTF-8 issues
 - Fix: Reduced file system access by the database
 - Fix: Fixed Microsoft Windows Media Player 12 support.
 - New: Initial support for audio recordings, i.e. recordings from radio
        stations. Live radio stations will follow very soon.
 - Fix: SortCriteria should work now.
 - New: Added some makefile tweaks.
        Thanks to Rolf Ahrenberg
 - New: Added Finnish translation.
        Thanks to Rolf Ahrenberg
 - New: Added hostname to friendlyName of the server.
        Thanks to Rolf Ahrenberg
 - New: Makefiles reviewed and almost completely rewritten to avoid
        redundancies.
        Thanks to Lucian Muresan
 - New: plugin makefile now links against the full file of the upnp plugin
        so-file. This makes symbolic link against the .so obsolete.
        Thanks to Lars Hanisch

2012-10-24: Version 1.0.0 RC1

 - New: Plug-ins for providing access to live-TV and recordings instead of
        integrated functions. This makes it easier to replace them. Additionally,
        new plug-ins may be integrated very easily.
 - New: The VDR plug-in live may be used for presentation home page.
 - New: The VDR plug-in streamdev is now used for live-TV. (via extern remux)
 - New: Tntnet is now used instead of the internal web server of libupnp. This
        enables the plugin to set HTTP headers without patching libupnp.
 - New: Live-TV channel icons are displayed as thumbnails in the folder preview.
 - New: Removed the requirement of FFMPEG in the basic plugins (live-TV and
        recordings). The plugin now uses only features of the VDR to detect
        the correct stream type of recordings.
 - New: Seeking is now finally possible on some devices (I have only tested
        Samsung TV-sets, yet). In this first attempt, it will use byte based
        seeks. Further attempts to improve it may be required.
 - New: Full rewrite of the rest of the plugin to reduce memory leaks as much
        as possible to me.

2009-01-25: Version 0.0.2-alpha1

 - Bug: A minor bug prevented the plugin from detecting the resources
        correnctly

2009-01-24: Version 0.0.2-alpha0

 - New: Added record streaming support for SDTV
        HDTV is not working yet, as it requires different DLNA profiles
 - Bug: Improved live-tv streaming quality, which will hopefully fix the
        problem with the buffer strategy and channels with bad bandwidth
 - New: Rearranged folder structure to meet future features
 - New: Added german translation
 - New: support for DLNA compatible devices. For those devices a special
        patch is required which must be applied to libupnp-1.6.6.
        See README for more information
 - New: Added new command line option to specify the destination folder
        for the metadata database and the webfolder.
        See README for more information

2009-11-19: Version 0.0.1-alpha4

- New: Added options for verbosity level and autodetect settings.

2009-10-23: Version 0.0.1-alpha1

- Fixed #185: Database rejected statements with single quotes inside strings

2009-10-23: Version 0.0.1-alpha0

- Initial revision.
- Known limitations: currently only channel lists with less than 30 channels
  work.
- LiveTV support
- Limited sort capabilities: Title, Creator, Write status, Publisher,
  Description, Long description, Genre
  Note: Sort will only work if this properties are present


dlna/dlna.cpp
/* 
 * File:   dlna.cpp
 * Author: savop
 * 
 * Created on 18. April 2009, 23:27
 */

#include <stdio.h>
#include <vdr/tools.h>
#include "upnp/dlna.h"

cDlna* cDlna::mInstance = NULL;

cDlna* cDlna::getInstance(void){
    if(cDlna::mInstance == NULL)
        cDlna::mInstance = new cDlna;

    if(cDlna::mInstance != NULL)
        return cDlna::mInstance;
    else return NULL;
}

cDlna::cDlna() {
    this->mRegisteredProfiles.clear();
    this->init();
}

cDlna::~cDlna() {
    this->mRegisteredProfiles.clear();
}


void cDlna::init(void){
    this->registerProfiles();
}

void cDlna::registerProfile(DLNAProfile* Profile){
    this->mRegisteredProfiles.push_back(Profile);
}

void cDlna::registerProfiles(){
    this->registerProfile(&DLNA_PROFILE_MPEG_TS_SD_EU);
    this->registerProfile(&DLNA_PROFILE_AVC_TS_HD_EU);
    this->registerProfile(&DLNA_PROFILE_MPEG_TS_SD_EU_ISO);
    this->registerProfile(&DLNA_PROFILE_AVC_TS_HD_EU_ISO);
}

const char* cDlna::getSupportedProtocols(){
    cString Protocols;
    list<DLNAProfile*>::iterator it;
    for(it=this->mRegisteredProfiles.begin(); it!=this->mRegisteredProfiles.end(); it++){
        Protocols = cString::sprintf("%s%s%s",(*Protocols)?*Protocols:"",(*Protocols)?",":"",this->getProtocolInfo(*it, DLNA_SUPPORTED_FLAGS));
    }
    return Protocols;
}

const char* cDlna::getProtocolInfo(DLNAProfile *Profile, int Op, const char* Ps, int Ci, unsigned int Flags){
    cString DLNA4thField = NULL;
    DLNA4thField = cString::sprintf("DLNA.ORG_PN=%s", Profile->ID);
    if(Op != -1)
        DLNA4thField = cString::sprintf("%s;DLNA.ORG_OP=%d",*DLNA4thField,Op);
    if(Ps != NULL)
        DLNA4thField = cString::sprintf("%s;DLNA.ORG_PS=%s",*DLNA4thField,Ps);
    if(Ci != -1)
        DLNA4thField = cString::sprintf("%s;DLNA.ORG_CI=%d",*DLNA4thField,Ci);
    if(Flags != 0)
        DLNA4thField = cString::sprintf("%s;DLNA.ORG_FLAGS=%.8x%.24x",*DLNA4thField,Flags,0);

    char* Protocol = strdup(cString::sprintf("http-get:*:%s:%s", Profile->mime, *DLNA4thField));
    return Protocol;
}

const char* cDlna::getDeviceDescription(const char* URLBase){
    cString description = cString::sprintf(
            "<?xml version = \"1.0\" encoding = \"utf-8\"?> \
            <root xmlns=\"%s\" xmlns:%s=\"%s\"> \
                <specVersion> \
                    <major>1</major> \
                    <minor>0</minor> \
                </specVersion> \
                <URLBase>%s</URLBase> \
                <device> \
                    <deviceType>%s</deviceType> \
                    <friendlyName>%s</friendlyName> \
                    <manufacturer>%s</manufacturer> \
                    <manufacturerURL>%s</manufacturerURL> \
                    <modelDescription>%s</modelDescription> \
                    <modelName>%s</modelName> \
                    <modelNumber>%s</modelNumber> \
                    <modelURL>%s</modelURL> \
                    <serialNumber>%s</serialNumber> \
                    <UDN>%s</UDN> \
                    <iconList> \
                        <icon> \
                            <mimetype>%s</mimetype> \
                            <width>%d</width> \
                            <height>%d</height> \
                            <depth>%d</depth> \
                            <url>%s</url> \
                        </icon> \
                        <icon> \
                            <mimetype>%s</mimetype> \
                            <width>%d</width> \
                            <height>%d</height> \
                            <depth>%d</depth> \
                            <url>%s</url> \
                        </icon> \
                        <icon> \
                            <mimetype>%s</mimetype> \
                            <width>%d</width> \
                            <height>%d</height> \
                            <depth>%d</depth> \
                            <url>%s</url> \
                        </icon> \
                        <icon> \
                            <mimetype>%s</mimetype> \
                            <width>%d</width> \
                            <height>%d</height> \
                            <depth>%d</depth> \
                            <url>%s</url> \
                        </icon> \
                    </iconList> \
                    <presentationURL>%s</presentationURL> \
                    <%s:X_DLNADOC>%s</dlna:X_DLNADOC> \
                    <serviceList> \
                        <service> \
                            <serviceType>%s</serviceType> \
                            <serviceId>%s</serviceId> \
                            <SCPDURL>%s</SCPDURL> \
                            <controlURL>%s</controlURL> \
                            <eventSubURL>%s</eventSubURL> \
                        </service> \
                        <service> \
                            <serviceType>%s</serviceType> \
                            <serviceId>%s</serviceId> \
                            <SCPDURL>%s</SCPDURL> \
                            <controlURL>%s</controlURL> \
                            <eventSubURL>%s</eventSubURL> \
                        </service> \
                    </serviceList> \
                </device> \
            </root>",
            UPNP_XMLNS_UPNP_DEV,            // UPnP Device Namespace (2)
            UPNP_XMLNS_PREFIX_DLNA,         // DLNA Namespace prefix (2)
            UPNP_XMLNS_DLNA_DEV,            // DLNA Device Namespace (2)
            URLBase,                        // URLBase (IP:PORT) (7)
            UPNP_DEVICE_TYPE,               // UPnP Device Type (MediaServer:1) (9)
            UPNP_DEVICE_FRIENDLY_NAME,      // UPnP Device Friendly Name (10)
            UPNP_DEVICE_MANUFACTURER,       // UPnP Device Manufacturer (11)
            UPNP_DEVICE_MANUFACTURER_URL,   // UPnP Device Manufacturer URL (12)
            UPNP_DEVICE_MODEL_DESCRIPTION,  // UPnP Device Model Description (13)
            UPNP_DEVICE_MODEL_NAME,         // UPnP Device Model Name (14)
            UPNP_DEVICE_MODEL_NUMBER,       // UPnP Device Model Number (15)
            UPNP_DEVICE_MODEL_URL,          // UPnP Device Model URL (16)
            UPNP_DEVICE_SERIAL_NUMBER,      // UPnP Device Serialnumber (17)
            UPNP_DEVICE_UDN,                // UPnP Device UDN (18)
            DLNA_ICON_JPEG_LRG_24.mime,     // UPnP Device Large Icon JPEG Mimetype (21)
            DLNA_ICON_JPEG_LRG_24.width,    // UPnP Device Large Icon Width (22)
            DLNA_ICON_JPEG_LRG_24.height,   // UPnP Device Large Icon Height (23)
            DLNA_ICON_JPEG_LRG_24.bitDepth, // UPnP Device Large Icon Bit Depth (24)
            UPNP_DEVICE_ICON_JPEG_LRG,      // UPnP Device Large Icon Path (25)
            DLNA_ICON_JPEG_SM_24.mime,      // UPnP Device Small Icon JPEG Mimetype (28)
            DLNA_ICON_JPEG_SM_24.width,     // UPnP Device Small Icon Width (29)
            DLNA_ICON_JPEG_SM_24.height,    // UPnP Device Small Icon Height (30)
            DLNA_ICON_JPEG_SM_24.bitDepth,  // UPnP Device Small Icon Bit Depth (31)
            UPNP_DEVICE_ICON_JPEG_SM,       // UPnP Device Small Icon Path (32)
            DLNA_ICON_PNG_SM_24A.mime,      // UPnP Device Small Icon PNG Mimetype (35)
            DLNA_ICON_PNG_SM_24A.width,     // UPnP Device Small Icon Width (36)
            DLNA_ICON_PNG_SM_24A.height,    // UPnP Device Small Icon Height (37)
            DLNA_ICON_PNG_SM_24A.bitDepth,  // UPnP Device Small Icon Bit Depth (38)
            UPNP_DEVICE_ICON_PNG_SM,        // UPnP Device Small Icon Path (39)
            DLNA_ICON_PNG_LRG_24A.mime,     // UPnP Device Large Icon PNG Mimetype (42)
            DLNA_ICON_PNG_LRG_24A.width,    // UPnP Device Large Icon Width (43)
            DLNA_ICON_PNG_LRG_24A.height,   // UPnP Device Large Icon Height (44)
            DLNA_ICON_PNG_LRG_24A.bitDepth, // UPnP Device Large Icon Bit Depth (45)
            UPNP_DEVICE_ICON_PNG_LRG,       // UPnP Device Large Icon Path (46)
            UPNP_WEB_PRESENTATION_URL,      // UPnP Presentation URL (49)
            UPNP_XMLNS_PREFIX_DLNA,         // DLNA Namespace prefix (50)
            DLNA_DEVICE_DMS_1_5,            // DLNA Device Type/Version (50)
            UPNP_CMS_SERVICE_TYPE,          // UPnP CMS Service Type
            UPNP_CMS_SERVICE_ID,            // UPnP CMS Service ID
            UPNP_CMS_SCPD_URL,              // UPnP CMS Service Description
            UPNP_CMS_CONTROL_URL,           // UPnP CMS Control URL
            UPNP_CMS_EVENT_URL,             // UPnP CMS Event URL
            UPNP_CDS_SERVICE_TYPE,          // UPnP CDS Service Type
            UPNP_CDS_SERVICE_ID,            // UPnP CDS Service ID
            UPNP_CDS_SCPD_URL,              // UPnP CDS Service Description
            UPNP_CDS_CONTROL_URL,           // UPnP CDS Control URL
            UPNP_CDS_EVENT_URL             // UPnP CDS Event URL
//            UPNP_AVT_SERVICE_TYPE,          // UPnP AVT Service Type
//            UPNP_AVT_SERVICE_ID,            // UPnP AVT Service ID
//            UPNP_AVT_SCPD_URL,              // UPnP AVT Service Description
//            UPNP_AVT_CONTROL_URL,           // UPnP AVT Control URL
//            UPNP_AVT_EVENT_URL              // UPnP AVT Event URL
            );
    return strdup0(*description);
}